File: frontend/bioreact/lib/model/bioreactor.dart
import 'package:bench_core/channels.dart';
import 'package:bench_core/mqtt.dart';
import 'dart:async';
import 'thermal_mass.dart';
import 'stirrer.dart';

class BioreactorModel {
  ThermalMassSystem thermal = ThermalMassSystem();
  StirrerModel stirrer = StirrerModel();

  void advanceTime(double tLast, double dt) {
    thermal.advanceTime(tLast, dt);
    stirrer.advanceTime(tLast, dt);
  }
}

class BioreactorBus extends ChannelBus {
  ThermalBus thermal;
  StirrerBus stirrer;

  BioreactorBus._({
    required this.thermal,
    required this.stirrer,
  });

  factory BioreactorBus() {
    var instance = BioreactorBus._(
      thermal: ThermalBus(),
      stirrer: StirrerBus(),
    );
    return instance;
  }

  @override
  String get busId => "bioreactor";

  @override
  List<ChannelBus> get children => [thermal, stirrer];
}

class BioreactorDeviceConnector extends DeviceConnector<BioreactorBus> {
  BioreactorModel device;
  ThermalDeviceConnector thermal;
  StirrerDeviceConnector stirrer;

  BioreactorDeviceConnector._(
    this.device, {
    required this.thermal,
    required this.stirrer,
  });

  factory BioreactorDeviceConnector(BioreactorModel device) {
    var thermalConnector = ThermalDeviceConnector(device.thermal);
    var stirrerConnector = StirrerDeviceConnector(device.stirrer);

    return BioreactorDeviceConnector._(
      device,
      thermal: thermalConnector,
      stirrer: stirrerConnector,
    );
  }

  factory BioreactorDeviceConnector.mockupMqtt(MqttService mqtt) {
    double currentTime = 0;
    // const double tEps = 1e-3;
    const double dt = 1.0;

    var bioreactor = BioreactorModel();
    var connector = BioreactorDeviceConnector(bioreactor);
    var deviceBus = BioreactorBus();
    connector.connectForwardChannels(deviceBus);
    mqtt.connectBus(deviceBus.deviceSideTree());
    connector.connectReverseChannels(deviceBus);

    Timer.periodic(
      Duration(milliseconds: (1000 * dt).toInt()),
      (timer) {
        bioreactor.advanceTime(currentTime, dt);
        currentTime += dt;
        print("t = $currentTime");
        connector.sample();
      },
    );

    return connector;
  }

  @override
  void connectForwardChannels(BioreactorBus bus) {
    thermal.connectForwardChannels(bus.thermal);
    stirrer.connectForwardChannels(bus.stirrer);
  }

  @override
  void connectReverseChannels(BioreactorBus bus) {
    thermal.connectReverseChannels(bus.thermal);
    stirrer.connectReverseChannels(bus.stirrer);
  }

  @override
  void sample() {
    thermal.sample();
    stirrer.sample();
  }
}

class BioreactorControlConnector extends ControlConnector<BioreactorBus> {
  ThermalControlConnector thermal;
  StirrerControlConnector stirrer;

  BioreactorControlConnector({required this.thermal, required this.stirrer});

  factory BioreactorControlConnector.empty() {
    return BioreactorControlConnector(
      stirrer: StirrerControlConnector(),
      thermal: ThermalControlConnector(),
    );
  }

  @override
  void connectForwardChannels(BioreactorBus bus) {
    thermal.connectForwardChannels(bus.thermal);
    stirrer.connectForwardChannels(bus.stirrer);
  }

  @override
  void connectReverseChannels(BioreactorBus bus) {
    thermal.connectReverseChannels(bus.thermal);
    stirrer.connectReverseChannels(bus.stirrer);
  }

  @override
  void dispose() {
    thermal.dispose();
    stirrer.dispose();
  }
}

File: backend/model/src/model/bioreactor.rs
use super::thermal::{ThermalMassModel, ThermalBus};
use super::stirrer::{StirrerModel, StirrerBus};
use flowmbed_peripherals::channels::{DeviceBusConnector, IOConnector};
use flowmbed_peripherals::util::QualifiedPath;
use flowmbed_peripherals::mqtt::MqttService;

pub struct BioreactorModel {
  thermal: ThermalMassModel,
  stirrer: StirrerModel,
}

impl BioreactorModel {
  pub fn new() -> Self {
    BioreactorModel { 
      thermal: ThermalMassModel::new(25.0),
      stirrer: StirrerModel::new() 
    }
  }

  pub fn step(&mut self, t_last: f32, dt: f32) {
    self.thermal.step(t_last, dt);
    self.stirrer.step(t_last, dt);
  }
}

pub struct BioreactorBus {
  id: String,
  thermal: ThermalBus,
  stirrer: StirrerBus,
}

impl BioreactorBus {
  pub fn new(id: &str) -> Self {
    BioreactorBus { 
      id: id.to_owned(),
      thermal: ThermalBus::new("thermal"),
      stirrer: StirrerBus::new("stirrer"),
    }
  }
}

impl DeviceBusConnector<BioreactorModel> for BioreactorBus {
  fn sample(&self, device: &BioreactorModel) {
    self.thermal.sample(&device.thermal);
    self.stirrer.sample(&device.stirrer);
  }
  fn handle_actions(&self, device: &mut BioreactorModel) {
    self.thermal.handle_actions(&mut device.thermal);
    self.stirrer.handle_actions(&mut device.stirrer);
    
  }
}

impl IOConnector for BioreactorBus {  
  fn connect_io(&mut self, comm: &mut dyn MqttService, qpath: &QualifiedPath) {
      self.thermal.connect_io(comm, &qpath.append(&self.id));
      self.stirrer.connect_io(comm, &qpath.append(&self.id));
  }
}

